ga/client/rttestimator.cpp

```diff
@@ -0,0 +1,207 @@
+	ga_error("rttestimator: Binding to socket\n");
+	if(bind(*sock, (struct sockaddr *) &servaddr, sizeof(servaddr)) < 0){
+		char *errmsg;
+		errmsg = strerror(errno);
+		ga_error("rttestimator: Failed to bind - %s\n", errmsg);
+		return -1;
+	}
+
+	return 0;
+}
+
+// Update the RTT array given the specified index and value.
+// Stores calculated response times returned from the UDP connection.
+void rtt_update(unsigned int index, unsigned int rtt_val) {
+	// Write to array
+	pthread_mutex_lock(&rtt_mutex);
+	rtt_store[index] = rtt_val;
+	pthread_mutex_unlock(&rtt_mutex);
+}
+
+// Primary thread for RTT estimation:
+// The client opens a connection with the server, and sends an RTT ctrlmsg.
+// It then listens for responses from the server.
+void *
+rttestimator_thread(void *param) {
+	in_addr ipaddr = *((in_addr *) param);
+	unsigned int recvlen;
+	struct timeval end;
+
+	// Initialize socket descriptor
+#ifdef _WIN32
+	SOCKET sock;
+#else
+	int sock;
+#endif
+
+	struct sockaddr_in servaddr;
+	// Set server information
+	memset(&servaddr, 0, sizeof(struct sockaddr_in));
+	servaddr.sin_family = AF_INET;
+	servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
+	servaddr.sin_port = htons(PORT);
+
+	if (rttestimator_init(ipaddr, (void *) &sock, servaddr)) {
+		return NULL;
+	}
+
+	// Send ctrlmsg to server to initiate server messages
+	ga_error("rttestimator_thread: Sending rttestimator ctrl message\n");
+	ctrlmsg_t msg;
+	ctrlsys_rttestimator(&msg);
+	ctrl_client_sendmsg(&msg, sizeof(ctrlmsg_system_rttestimator_t));
+
+	ping_id_iterator = 0;
+	// Zero rtt_store
+	for (int i = 0; i < RTT_STORE_SIZE; i++) {
+		rtt_store[i] = 0;
+	}
+	pthread_mutex_init(&rtt_mutex, NULL);
+
+	char *buf;
+	buf = (char *) malloc(BUFSIZE);
+
+	// Loop forever
+	while(1){
+		// Wait to receive a packet
+		recvlen = recvfrom(sock, buf, BUFSIZE, 0, NULL, NULL);
+
+		if (recvlen >= sizeof(bbr_rtt_t)) {
+			char *read_head = buf;
+			bbr_rtt_t read_data;
+
+			while (read_head - buf < recvlen) {
+				memcpy(&read_data, read_head, sizeof(bbr_rtt_t));
+				// Record time of value and compute RTT
+				gettimeofday(&end, NULL);
+				unsigned int rtt_val = (end.tv_sec - read_data.time_record.tv_sec) * 1000000.0 + 
+					end.tv_usec - read_data.time_record.tv_usec;
+				
+				// NOTE: This does not take into account dropped packets or skipped replies.
+				rtt_update(read_data.rtt_id, rtt_val);
+
+				read_head += sizeof(bbr_rtt_t);
+			}
+		}
+	}
+	free(buf);
+
+// Clean up sockets and shut down the connection
+#ifdef _WIN32
+	WSACleanup();
+#endif
+
+	int status = 0;
+#ifdef _WIN32
+	status = shutdown(sock, SD_BOTH);
+	if(status == 0){ status = closesocket(sock); }
+#else
+	status = shutdown(sock, SHUT_RDWR);
+	if(status == 0){ status = close(sock); }
+#endif
+
+	return NULL;
+}
+
+unsigned int getRtprop() {
+	unsigned int ret = UINT_MAX;
+	int rtprop_window_size = RTPROP_WINDOW_SIZE * 1000000 / PING_DELAY;
+	
+	pthread_mutex_lock(&rtt_mutex);
+	for (unsigned int i = ping_id_iterator; 
```

Comment:
i will always be 0 because ping_id_iterator is no longer used in rttestimator_thread